boa/src/builtins/map/mod.rs

```diff
@@ -419,79 +441,62 @@ impl Map {
         args: &[JsValue],
         context: &mut Context,
     ) -> JsResult<JsValue> {
-        if args.is_empty() {
-            return Err(JsValue::new("Missing argument for Map.prototype.forEach"));
-        }
-
-        let callback_arg = &args[0];
+        // 1. Let M be the this value.
+        // 2. Perform ? RequireInternalSlot(M, [[MapData]]).
+        let map = match this {
+            JsValue::Object(obj) if obj.is_map() => obj,
+            _ => return context.throw_type_error("`this` is not a Map"),
+        };
 
-        if !callback_arg.is_function() {
-            let name = callback_arg.to_string(context)?;
-            return context.throw_type_error(format!("{} is not a function", name));
-        }
+        // 3. If IsCallable(callbackfn) is false, throw a TypeError exception.
+        let callback = match args.get_or_undefined(0) {
+            JsValue::Object(obj) if obj.is_function() => obj,
+            val => {
+                let name = val.to_string(context)?;
+                return context.throw_type_error(format!("{} is not a function", name));
+            }
+        };
 
         let this_arg = args.get_or_undefined(1);
 
         let mut index = 0;
 
-        let lock = Map::lock(this, context)?;
-
-        while index < Map::get_full_len(this, context)? {
-            let arguments = if let JsValue::Object(ref object) = this {
-                let object = object.borrow();
-                if let Some(map) = object.as_map_ref() {
-                    if let Some((key, value)) = map.get_index(index) {
-                        Some([value.clone(), key.clone(), this.clone()])
-                    } else {
-                        None
-                    }
+        // NOTE:
+        //
+        // forEach does not directly mutate the object on which it is called but
+        // the object may be mutated by the calls to callbackfn. Each entry of a
+        // map's [[MapData]] is only visited once. New keys added after the call
+        // to forEach begins are visited. A key will be revisited if it is deleted
+        // after it has been visited and then re-added before the forEach call completes.
+        // Keys that are deleted after the call to forEach begins and before being visited
+        // are not visited unless the key is added again before the forEach call completes.
+        let lock = map.borrow_mut().expect_map_mut().lock(map.clone());
```

Comment:
We can prefix the `lock` variable with `_` and avoid the manual `drop` later.
```suggestion
        let _lock = map.borrow_mut().expect_map_mut().lock(map.clone());
```